src/main/java/com/google/devtools/build/remote/client/LogPrintingUtils.java

```diff
@@ -0,0 +1,144 @@
+import java.util.HashMap;
+import java.util.Map;
+
+/** Methods for printing log files. */
+public class LogPrintingUtils {
+  private static final String DELIMETER =
+      "---------------------------------------------------------\n";
+
+  /**
+   * Attempt to find and print ExecuteResponse from the details of a log entry for a Watch call. If
+   * no Operation could be found in the Watch call responses, or an Operation was found but failed,
+   * a failure message is printed.
+   */
+  private static void printExecuteResponse(WatchDetails watch) throws IOException {
+    for (ChangeBatch cb : watch.getResponsesList()) {
+      for (Change ch : cb.getChangesList()) {
+        if (ch.getState() != State.EXISTS) {
+          continue;
+        }
+        Operation o = ch.getData().unpack(Operation.class);
+        if (o.getResultCase() == ResultCase.ERROR && o.getError().getCode() != Code.OK.value()) {
+          System.out.printf("Error status in ChangeBatches: %s\n", o.getError().toString());
+          return;
+        } else if (o.getResultCase() == ResultCase.RESPONSE && o.getDone()) {
+          System.out.println("ExecuteResponse extracted:");
+          System.out.println(o.getResponse().unpack(ExecuteResponse.class).toString());
+          return;
+        }
+      }
+    }
+    System.out.println("Could not find ExecuteResponse in Watch call details.");
+  }
+
+  /** Print an individual log entry. */
+  private static void printLogEntry(LogEntry entry) throws IOException {
+    System.out.println(entry.toString());
+    if (entry.getDetails().getDetailsCase() == DetailsCase.WATCH) {
+      System.out.println("\nAttempted to extract ExecuteResponse from Watch call responses:");
+      printExecuteResponse(entry.getDetails().getWatch());
+    }
+  }
+
+  /**
+   * Prints each entry out individually (ungrouped) and a message at the end for how many entries
+   * were printed/skipped.
+   */
+  private static void printEntriesInOrder(PrintLogCommand options) throws IOException {
+    try (InputStream in = new FileInputStream(options.file)) {
+      LogEntry entry;
+      while ((entry = LogEntry.parseDelimitedFrom(in)) != null) {
+        printLogEntry(entry);
+        System.out.print(DELIMETER);
+      }
+    }
+  }
+
+  /**
+   * Gives comparator values for timestamps.
+   *
+   * @return {@code 0} if {@code t1} is equal to {@code t2}; a value less than {code 0} if {@code
+   *     t1} occurs before {@code t2}; a value greater than {code 0} if{@code t1} occurs after
+   *     {@code t2}.
+   */
+  private static int compareTimestamps(Timestamp t1, Timestamp t2) {
+    int cmpSeconds = Long.compare(t1.getSeconds(), t2.getSeconds());
+    return (cmpSeconds != 0) ? cmpSeconds : Integer.compare(t1.getNanos(), t2.getNanos());
+  }
+
+  /**
+   * Prints each entry in groups by action, and a message at the end for how many entries were
+   * printed/skipped. If an entry does not have metadata to identify the action, is it skipped.
```

Comment:
If there's non-zero of those, would you mind printing a message (e.g., "X lines missing action information")?

Reply:
Done